src/Sulu/Bundle/MediaBundle/Collection/Manager/CollectionManager.php

```diff
@@ -797,4 +846,277 @@ private function getCollectionMeta(CollectionInterface $collection, ?string $loc
 
         return $meta;
     }
+
+    /**
+     * @param array<array{id: int, resourceKey: string, depth: int}> $resources
+     *
+     * @return array<int, array<array{id: int, resourceKey: string}>>
+     */
+    private function groupResourcesByDepth(array $resources)
+    {
+        $grouped = [];
+
+        foreach ($resources as $resource) {
+            $depth = $resource['depth'];
+            unset($resource['depth']);
+
+            if (!isset($grouped[$depth])) {
+                $grouped[$depth] = [];
+            }
+
+            $grouped[$depth][] = $resource;
+        }
+
+        \krsort($grouped);
+
+        return \array_values($grouped);
+    }
+
+    /**
+     * @param array<array{id: int, resourceKey: string, depth: int}> $resources
+     *
+     * @return array<string, array{id: int, resourceKey: string, depth: int}>
+     */
+    private function groupResourcesByKey(array $resources): array
+    {
+        $result = [];
+        foreach ($resources as $resource) {
+            $resourceKey = $resource['resourceKey'];
+            $id = (string) $resource['id'];
+
+            $result[$resourceKey . '-' . $id] = $resource;
+        }
+
+        return $result;
+    }
+
+    private function createChildCollectionsQueryBuilder(int $id): QueryBuilder
+    {
+        return $this->em->createQueryBuilder()
+            ->select('collection.id AS id')
+            ->addSelect('\'collections\' AS resourceKey')
+            ->addSelect('collection.depth as depth')
+            ->from(CollectionInterface::class, 'collection')
+            ->leftJoin(CollectionInterface::class, 'parent', Join::WITH, 'collection.lft > parent.lft AND collection.rgt < parent.rgt')
+            ->where('parent.id = :id')
+            ->setParameter('id', $id);
+    }
+
+    /**
+     * @return array<array{id: int, resourceKey: string, depth: int}>
+     */
+    private function findChildCollections(int $id): array
```

Comment:
Best would be that all methods which communicate with the database live in the `CollectionRepository`. The `Repository` should be in the future architecture the only service creating this kind of queries.

We need to discuss here the naming as it are queries for a specific usecase which are not extendable queries.

I think here about `findChildCollectionIdsForPermissionCheck` or `findChildCollectionIdsForDeletion`. /cc @sulu/core-team 

Another way would also that we use the `findFlat($filters, $fields)` method which should use the field descriptors and so we could there add the required filters e.g.: `findFlat(['not-permission' => 'delete', 'rootCollectionId' => $id], ['id', 'depth'])` and `findFlat(['rootCollectionId' => $id], ['id', 'depth'])`.

Reply:
Didn't implement this suggestion for now, until we have a decision on how to do it

Reply:
should be fixed now